### PR TITLE
Zenjectified it all

### PR DESCRIPTION
**Pull request by ErisApps**
The PR mainly aims change the mod the DynamicInit and to rework the architecture to leverage Zenject.
As a side-effect, this PR also fixes issue #2

### CODE REVIEW

File: NsvAssetLoader.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace SliceVisualizer
+{
+    internal class NsvAssetLoader : IInitializable, IDisposable
+    {
+        private readonly SiraLog _siraLog;
+
+        private Material? _uiNoGlowMaterial;
+
+        public Material UINoGlowMaterial => _uiNoGlowMaterial ??= new Material(Resources.FindObjectsOfTypeAll<Shader>().First(x => x.name == "Custom/Sprite"));
```
Comment (m1el) [inline NsvAssetLoader.cs:16]: 👍

File: Core/NsvController.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+using UnityEngine.UI;
+using Zenject;
+
+namespace SliceVisualizer.Core
+{
+    internal class NsvController : IInitializable, ITickable, IDisposable
+    {
+        private readonly PluginConfig _config;
+        private readonly BeatmapObjectManager _beatmapObjectManager;
+        private readonly NsvSlicedBlock.Pool _slicedBlockPool;
```
Comment (m1el) [inline Core/NsvController.cs]: OK, please hear me out on this: using a memory pool here is less efficient than having a static pool of objects.
Spawning/Despawning/memory management will take more time during the game than just spawning these 12 objects at the start of the level.

File: Core/NsvSlicedBlock.cs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        private Color _arrowColor;
+        private Color _missedAreaColor;
+        private Color _sliceColor;
+        private bool _needsUpdate;
+
+        [Inject]
+        internal void Construct(PluginConfig config, NsvAssetLoader assetLoader, ColorManager colorManager)
+        {
+            _colorManager = colorManager;
+            _assetLoader = assetLoader;
```
Comment (m1el) [inline Core/NsvSlicedBlock.cs]: `_colorManager` and `_assetLoader` are only required during build process, so no need to hold reference to those after constructing. So I suggest either not storing them or `null`ing them

File: Core/NsvSlicedBlock.cs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+            }
+
+            _aliveTime = 0f;
+
+            _isDirectional = noteData.cutDirection != NoteCutDirection.Any && noteData.cutDirection != NoteCutDirection.None;
+
+            if (_config.UseCustomColors)
+            {
+                _color = noteData.colorType == ColorType.ColorA ? _config.LeftColor : _config.RightColor;
+                _otherColor = noteData.colorType != ColorType.ColorA ? _config.LeftColor : _config.RightColor;
```
Comment (m1el) [inline Core/NsvSlicedBlock.cs]: Can this be a single conditional with more assignments please?

File: Core/NsvSlicedBlock.cs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+            }
+            else
+            {
+                // This should work due to colors of both type A and B match for their respective saber and blocks
+                _color = _colorManager.ColorForType(noteData.colorType);
+                _otherColor = _colorManager.ColorForSaberType(noteCutInfo.saberType);
```
Comment (m1el) [inline Core/NsvSlicedBlock.cs]: I think it makes sense to rename this to saberColor

**Review by m1el — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,357 @@
-.vs
-bin
-obj
+## Ignore Visual Studio temporary files, build results, and
```
Comment (m1el) [inline .gitignore]: Can we not do this?
This `.gitignore` seems too much.
There are c++, python, javascript, tfs ignores here, and nothing of that ever going to be used here.

**Comment by m1el** [comment]
Please also update the readme since new dependency requires extra steps for installation.

File: Core/NsvSlicedBlock.cs (hunk 4 above)
Comment (m1el) [reply Core/NsvSlicedBlock.cs]: the `saberColor` comment also applies here. so probably not one conditional, but different conditions.

**Comment by ErisApps** [comment]
Totally forgot about the README, will change that as well.

File: Core/NsvSlicedBlock.cs (hunk 3 above)
Comment (ErisApps) [reply Core/NsvSlicedBlock.cs]: Will pass the assetLoader reference, as a parameter instead. However, the _colorManager is used multiple times later on, so I'll keep that one as is.

File: Core/NsvSlicedBlock.cs (hunk 5 above)
Comment (ErisApps) [reply Core/NsvSlicedBlock.cs]: Will change

File: Core/NsvSlicedBlock.cs (hunk 4 above)
Comment (ErisApps) [reply Core/NsvSlicedBlock.cs]: Could you elaborate on this one?

Comment (m1el) [reply Core/NsvSlicedBlock.cs]: 1. rename `_otherColor` to `_saberColor`.
2. `_saberColor = noteCutInfo.saberType == ColorType.ColorA ? ...`

File: README.md [hunk 7]
```diff
@@ -41,7 +41,7 @@ Set to 0.0 to disable this effect.
 - `CenterScale` (default: 0.2) -- scale of the circle in the center, relative to note scale.
 - `ArrowScale` (default: 0.6) -- scale of the note arrow, relative to note scale.
 - `UIOpacity` (default: 1.0) -- Overall opacity for the UI.  Multiplicative with opacity/alpha of other colors.
-- `CanvasOffset` (default: `[0.0, 0.0, 16.0]`) -- control `[x, y, z]` position of bottom center point of the visualization canvas.  
+- `CanvasOffset` (default: `[0.0, 0.0, 16.0]`) -- control `[x, y, z]` position of bottom center point of the visualization canvas.
```
Comment (m1el) [inline README.md]: Two trailing spaces here are intended

Comment (ErisApps) [reply README.md]: I don't recall touching that :|
Will revert

File: Core/NsvSlicedBlock.cs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+            }
+
+            _aliveTime = 0f;
+
+            _isDirectional = noteData.cutDirection != NoteCutDirection.Any && noteData.cutDirection != NoteCutDirection.None;
+
+            if (_config.UseCustomColors)
+            {
+                _color = noteData.colorType == ColorType.ColorA ? _config.LeftColor : _config.RightColor;
+                _saberColor = noteData.colorType != ColorType.ColorA ? _config.LeftColor : _config.RightColor;
```
Comment (m1el) [inline Core/NsvSlicedBlock.cs]:
```suggestion
                _saberColor = noteCutInfo.saberType == SaberType.SaberA ? _config.LeftColor : _config.RightColor;
```

Comment (ErisApps) [reply Core/NsvSlicedBlock.cs]: Won't compile... but I get the point.
That being said, I'm actually not sure how this will react to left-handed mode. 🤔

Comment (m1el) [reply Core/NsvSlicedBlock.cs]: (updated the code suggestion to actual code that works)
I just checked it -- left-handed mode does two things:
1. mirror the notes and swap their color
2. use left saber in single saber mode

So this change works as intended.

**Review by m1el — APPROVED** [review]

**Comment by m1el** [comment]
Thanks for the contribution!

**Pull request merged**